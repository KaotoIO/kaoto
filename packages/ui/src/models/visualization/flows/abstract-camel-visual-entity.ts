/* eslint-disable no-case-declarations */
import { ProcessorDefinition } from '@kaoto/camel-catalog/types';
import { SchemaService } from '../../../components/Form/schema.service';
import { ROOT_PATH, getArrayProperty, getValue, setValue } from '../../../utils';
import { NodeIconResolver, NodeIconType } from '../../../utils/node-icon-resolver';
import { DefinedComponent } from '../../camel-catalog-index';
import { EntityType } from '../../camel/entities';
import {
  AddStepMode,
  BaseVisualCamelEntity,
  IVisualizationNode,
  IVisualizationNodeData,
  NodeInteraction,
  VisualComponentSchema,
} from '../base-visual-entity';
import { createVisualizationNode } from '../visualization-node';
import { CamelComponentDefaultService } from './support/camel-component-default.service';
import { CamelComponentSchemaService } from './support/camel-component-schema.service';
import { CamelProcessorStepsProperties, CamelRouteVisualEntityData } from './support/camel-component-types';
import { CamelStepsService } from './support/camel-steps.service';
import { ModelValidationService } from './support/validators/model-validation.service';

export abstract class AbstractCamelVisualEntity<T extends object> implements BaseVisualCamelEntity {
  constructor(public route: T) {}

  abstract id: string;
  abstract type: EntityType;
  abstract setId(id: string): void;
  abstract toJSON(): unknown;
  protected abstract getRootUri(): string | undefined;

  getId(): string {
    return this.id;
  }

  getNodeLabel(path?: string, labelType?: string): string {
    if (!path) return '';

    const componentModel = getValue(this.route, path);
    const label = CamelComponentSchemaService.getNodeLabel(
      CamelComponentSchemaService.getCamelComponentLookup(path, componentModel),
      componentModel,
      labelType,
    );

    return label;
  }

  getTooltipContent(path?: string): string {
    if (!path) return '';
    const componentModel = getValue(this.route, path);

    const content = CamelComponentSchemaService.getTooltipContent(
      CamelComponentSchemaService.getCamelComponentLookup(path, componentModel),
    );

    return content;
  }

  getComponentSchema(path?: string): VisualComponentSchema | undefined {
    if (!path) return undefined;

    const componentModel = getValue(this.route, path);
    const visualComponentSchema = CamelComponentSchemaService.getVisualComponentSchema(path, componentModel);

    return visualComponentSchema;
  }

  getOmitFormFields(): string[] {
    return SchemaService.OMIT_FORM_FIELDS;
  }

  updateModel(path: string | undefined, value: unknown): void {
    if (!path) return;
    const updatedValue = CamelComponentSchemaService.getUriSerializedDefinition(path, value);

    setValue(this.route, path, updatedValue);
  }

  /**
   * Add a step to the route
   *
   * path examples:
   *      from
   *      from.steps.0.setHeader
   *      from.steps.1.choice.when.0
   *      from.steps.1.choice.when.0.steps.0.setHeader
   *      from.steps.1.choice.otherwise
   *      from.steps.1.choice.otherwise.steps.0.setHeader
   *      from.steps.2.doTry.doCatch.0
   *      from.steps.2.doTry.doCatch.0.steps.0.setHeader
   */
  addStep(options: {
    definedComponent: DefinedComponent;
    mode: AddStepMode;
    data: IVisualizationNodeData;
    targetProperty?: string;
  }) {
    if (options.data.path === undefined) return;
    const defaultValue = CamelComponentDefaultService.getDefaultNodeDefinitionValue(options.definedComponent);
    const stepsProperties = CamelComponentSchemaService.getProcessorStepsProperties(
      (options.data as CamelRouteVisualEntityData).processorName as keyof ProcessorDefinition,
    );

    if (options.mode === AddStepMode.InsertChildStep || options.mode === AddStepMode.InsertSpecialChildStep) {
      this.insertChildStep(options, stepsProperties, defaultValue);
      return;
    }

    const pathArray = options.data.path.split('.');
    const last = pathArray[pathArray.length - 1];
    const penultimate = pathArray[pathArray.length - 2];

    /**
     * If the last segment is a string and the penultimate is a number, it means the target is member of an array
     * therefore we need to look for the array and insert the element at the given index + 1
     *
     * f.i. from.steps.0.setHeader
     * penultimate: 0
     * last: setHeader
     */
    if (!Number.isInteger(Number(last)) && Number.isInteger(Number(penultimate))) {
      /** If we're in Append mode, we need to insert the step after the selected index hence `Number(penultimate) + 1` */
      const desiredStartIndex = options.mode === AddStepMode.AppendStep ? Number(penultimate) + 1 : Number(penultimate);

      /** If we're in Replace mode, we need to delete the existing step */
      const deleteCount = options.mode === AddStepMode.ReplaceStep ? 1 : 0;

      const stepsArray: ProcessorDefinition[] = getValue(this.route, pathArray.slice(0, -2), []);
      stepsArray.splice(desiredStartIndex, deleteCount, defaultValue);

      return;
    }
  }

  removeStep(path?: string): void {
    if (!path) return;
    const pathArray = path.split('.');
    const last = pathArray[pathArray.length - 1];
    const penultimate = pathArray[pathArray.length - 2];

    /**
     * If the last segment is a number, it means the target object is a member of an array
     * therefore we need to look for the array and remove the element at the given index
     *
     * f.i. from.steps.1.choice.when.0
     * last: 0
     */
    let array = getValue(this.route, pathArray.slice(0, -1), []);
    if (Number.isInteger(Number(last)) && Array.isArray(array)) {
      array.splice(Number(last), 1);

      return;
    }

    /**
     * If the last segment is a word and the penultimate is a number, it means the target is an object
     * potentially a Processor, that belongs to an array, therefore we remove it entirely
     *
     * f.i. from.steps.1.choice
     * last: choice
     * penultimate: 1
     */
    array = getValue(this.route, pathArray.slice(0, -2), []);
    if (!Number.isInteger(Number(last)) && Number.isInteger(Number(penultimate)) && Array.isArray(array)) {
      array.splice(Number(penultimate), 1);

      return;
    }

    /**
     * If both the last and penultimate segment are words, it means the target is a property of an object
     * therefore we delete it
     *
     * f.i. from.steps.1.choice.otherwise
     * last: otherwise
     * penultimate: choice
     */
    const object = getValue(this.route, pathArray.slice(0, -1), {});
    if (!Number.isInteger(Number(last)) && !Number.isInteger(Number(penultimate)) && typeof object === 'object') {
      delete object[last];
    }
  }

  getNodeInteraction(data: IVisualizationNodeData): NodeInteraction {
    const processorName = (data as CamelRouteVisualEntityData).processorName;
    const canHavePreviousStep = CamelComponentSchemaService.canHavePreviousStep(processorName);
    const stepsProperties = CamelComponentSchemaService.getProcessorStepsProperties(processorName);
    const canHaveChildren = stepsProperties.find((property) => property.type === 'branch') !== undefined;
    const canHaveSpecialChildren = Object.keys(stepsProperties).length > 1;
    const canReplaceStep = CamelComponentSchemaService.canReplaceStep(processorName);
    const canRemoveStep = processorName !== ('from' as keyof ProcessorDefinition);
    const canRemoveFlow = data.path === ROOT_PATH;
    const canBeDisabled = CamelComponentSchemaService.canBeDisabled(processorName);

    return {
      canHavePreviousStep,
      canHaveNextStep: canHavePreviousStep,
      canHaveChildren,
      canHaveSpecialChildren,
      canReplaceStep,
      canRemoveStep,
      canRemoveFlow,
      canBeDisabled,
    };
  }

  getNodeValidationText(path?: string | undefined): string | undefined {
    const componentVisualSchema = this.getComponentSchema(path);
    if (!componentVisualSchema) return undefined;

    return ModelValidationService.validateNodeStatus(componentVisualSchema);
  }

  toVizNode(): IVisualizationNode {
    const routeGroupNode = createVisualizationNode(this.id, {
      path: ROOT_PATH,
      entity: this,
      isGroup: true,
      icon: NodeIconResolver.getIcon(this.type, NodeIconType.VisualEntity),
    });

    const fromNode = CamelStepsService.getVizNodeFromProcessor(
      'from',
      {
        processorName: 'from' as keyof ProcessorDefinition,
        componentName: CamelComponentSchemaService.getComponentNameFromUri(this.getRootUri()!),
      },
      this.route,
    );

    if (!this.getRootUri()) {
      fromNode.data.icon = NodeIconResolver.getPlaceholderIcon();
    }
    routeGroupNode.addChild(fromNode);

    return routeGroupNode;
  }

  private insertChildStep(
    options: Parameters<AbstractCamelVisualEntity<object>['addStep']>[0],
    stepsProperties: CamelProcessorStepsProperties[],
    defaultValue: ProcessorDefinition = {},
  ) {
    const property = stepsProperties.find((property) =>
      options.mode === AddStepMode.InsertChildStep ? 'steps' : options.definedComponent.name === property.name,
    );
    if (property === undefined) return;

    if (property.type === 'single-clause') {
      setValue(this.route, `${options.data.path}.${property.name}`, defaultValue);
    } else {
      const arrayPath = getArrayProperty(this.route, `${options.data.path}.${property.name}`);
      arrayPath.unshift(defaultValue);
    }
  }
}
