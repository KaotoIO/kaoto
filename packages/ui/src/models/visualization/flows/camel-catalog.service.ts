import { ComponentsCatalog } from '../../camel-catalog-index';
import { ICamelComponentDefinition } from '../../camel-components-catalog';
import { ICamelProcessorDefinition } from '../../camel-processors-catalog';
import { CatalogKind } from '../../catalog-kind';
import { IKameletDefinition } from '../../kamelets-catalog';

export class CamelCatalogService {
  private static catalogs: ComponentsCatalog = {};

  static getCatalogByKey<CATALOG_KEY extends CatalogKind>(catalogKey: CATALOG_KEY): ComponentsCatalog[CATALOG_KEY] {
    return this.catalogs[catalogKey];
  }

  static setCatalogKey<CATALOG_KEY extends CatalogKind>(
    catalogKey: CATALOG_KEY,
    catalog?: ComponentsCatalog[CATALOG_KEY],
  ): void {
    this.catalogs[catalogKey] = catalog;
  }

  static getComponent(catalogKey: CatalogKind.Component, componentName?: string): ICamelComponentDefinition | undefined;
  static getComponent(catalogKey: CatalogKind.Processor, componentName?: string): ICamelProcessorDefinition | undefined;
  static getComponent(catalogKey: CatalogKind.Kamelet, componentName?: string): IKameletDefinition | undefined;
  static getComponent(
    catalogKey: CatalogKind,
    componentName?: string,
  ): ICamelComponentDefinition | ICamelProcessorDefinition | IKameletDefinition | undefined {
    if (componentName === undefined) return undefined;

    return this.catalogs[catalogKey]?.[componentName];
  }

  /**
   * Public only as a convenience method for test
   * not meant to be used in production code
   */
  static clearCatalogs(): void {
    this.catalogs = {};
  }
}
