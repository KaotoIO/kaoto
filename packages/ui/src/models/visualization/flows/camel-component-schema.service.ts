import type { JSONSchemaType } from 'ajv';
import { useCatalogStore } from '../../../store';
import { ICamelProcessorProperty } from '../../camel-processors-catalog';
import { CatalogKind } from '../../catalog-kind';
import { VisualComponentSchema } from '../base-visual-entity';
import { ICamelComponentProperty } from '../../camel-components-catalog';
import { isDefined } from '../../../utils';

interface ICamelElementLookupResult {
  processorName: string;
  componentName?: string;
}

export class CamelComponentSchemaService {
  static getVisualComponentSchema(path: string, definition: any): VisualComponentSchema | undefined {
    const camelElementLookup = this.getCamelComponentLookup(path, definition);

    return {
      title: camelElementLookup.processorName,
      schema: this.getSchema(camelElementLookup),
      definition,
    };
  }

  static getCamelComponentLookup(path: string, definition: any): ICamelElementLookupResult {
    const splitPath = path.split('.');
    const lastPathSegment = splitPath[splitPath.length - 1];
    const pathAsIndex = Number.parseInt(lastPathSegment, 10);

    /**
     * If the last path segment is NaN, it means this is a Camel Processor
     * for instance, `from`, `when`, `otherwise` or `to` properties in a Route
     * and we can just return the path as the name of the component
     */
    if (Number.isNaN(pathAsIndex)) {
      return this.getCamelElement(lastPathSegment, definition);
    }

    /**
     * The last path segment is a number, it means is an array of objects
     * and we need to look for the previous path segment to get the name of the processor
     * for instance, a `when` property in a `Choice` processor
     */
    const previousPathSegment = splitPath[splitPath.length - 2];
    if (typeof previousPathSegment === 'string') {
      return this.getCamelElement(previousPathSegment, definition);
    }

    /**
     * If we reach this point, it means we couldn't determine the name of the component
     */
    return { processorName: '' };
  }

  /**
   * If the processor is a `from` or `to` processor, we need to extract the component name from the uri property
   * and return both the processor name and the underlying component name to build the combined schema
   */
  private static getCamelElement(processorName: string, definition: any): ICamelElementLookupResult {
    switch (processorName) {
      case 'from':
        return {
          processorName,
          componentName: this.getComponentNameFromUri(definition.uri),
        };

      case 'to':
        /** The To processor is using `to: timer:tick?period=1000` form */
        if (typeof definition === 'string') {
          return {
            processorName,
            componentName: this.getComponentNameFromUri(definition),
          };
        }

        /** The To processor is using `to: { uri: 'timer:tick?period=1000' }` form */
        return {
          processorName,
          componentName: this.getComponentNameFromUri(definition.uri),
        };

      default:
        return { processorName };
    }
  }

  /**
   * Extract the component name from the endpoint uri
   * An URI is composed by a component name and query parameters, separated by a colon
   * For instance: `timer:tick?period=1000`
   */
  private static getComponentNameFromUri(uri: string): string | undefined {
    if (!uri) {
      return undefined;
    }
    const uriParts = uri.split(':');

    return uriParts[0];
  }

  private static getSchema(camelElementLookup: ICamelElementLookupResult): JSONSchemaType<unknown> {
    const processorDefinition =
      useCatalogStore.getState().catalogs[CatalogKind.Processor]?.[camelElementLookup.processorName];

    if (processorDefinition === undefined) return {} as unknown as JSONSchemaType<unknown>;

    const schema = this.getSchemaFromCamelCommonProperties(processorDefinition.properties);

    if (camelElementLookup.componentName !== undefined) {
      const componentDefinition =
        useCatalogStore.getState().catalogs[CatalogKind.Component]?.[camelElementLookup.componentName];

      if (componentDefinition !== undefined) {
        const componentSchema = this.getSchemaFromCamelCommonProperties(componentDefinition.properties);
        schema.properties.parameters = {
          type: 'object',
          title: 'Component Parameters',
          description: 'Component parameters description',
          properties: componentSchema.properties,
        };
      }
    }

    return schema;
  }

  static getIconName(camelElementLookup: ICamelElementLookupResult): string | undefined {
    if (
      isDefined(camelElementLookup.componentName) &&
      isDefined(useCatalogStore.getState().catalogs[CatalogKind.Component]?.[camelElementLookup.componentName])
    ) {
      return camelElementLookup.componentName;
    }
    if (
      isDefined(camelElementLookup.processorName) &&
      !isDefined(camelElementLookup.componentName) &&
      isDefined(useCatalogStore.getState().catalogs[CatalogKind.Processor]?.[camelElementLookup.processorName])
    ) {
      return camelElementLookup.processorName;
    }
    return '';
  }

  /**
   * Transform Camel Common properties into a JSON Schema
   */
  private static getSchemaFromCamelCommonProperties(
    properties: Record<string, ICamelProcessorProperty | ICamelComponentProperty>,
  ): JSONSchemaType<unknown> {
    const required: string[] = [];
    const schema = {
      type: 'object',
      properties: {},
      required,
    } as unknown as JSONSchemaType<unknown>;

    Object.keys(properties).forEach((propertyName) => {
      const property = properties[propertyName];
      const propertyType = this.getJSONType(property);
      const propertySchema = {
        type: propertyType,
        title: property.displayName,
        description: property.description,
        deprecated: property.deprecated,
      } as unknown as JSONSchemaType<unknown>;

      if (property.enum !== undefined) {
        propertySchema.enum = property.enum;
      }

      if (property.required) {
        required.push(propertyName);
      }

      schema.properties[propertyName] = propertySchema;
    });

    return schema;
  }

  /**
   * Transform Camel property types into JSON Schema types
   *
   * This is needed because the Camel Catalog is using different types than JSON Schema
   * For instance, the Camel Catalog is using `duration` instead of `number`
   */
  static getJSONType(property: ICamelProcessorProperty | ICamelComponentProperty): string | undefined {
    /** Camel defines enum as a type, whereas it should be string and let uniforms handle the right field */
    if (Array.isArray(property.enum)) {
      return undefined;
    }

    switch (property.type) {
      /** Camel defines duration as string since it supports placeholders */
      case 'duration':
        return 'string';

      default:
        return property.type;
    }
  }
}
