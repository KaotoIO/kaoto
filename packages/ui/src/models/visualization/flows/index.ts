export * from './camel-catalog.service';
export * from './pipe-visual-entity';
export * from './kamelet-binding-visual-entity';
export * from './kamelet-visual-entity';
export * from './camel-route-visual-entity';
