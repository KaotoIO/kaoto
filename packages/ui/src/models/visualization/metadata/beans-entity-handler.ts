import { BeansAwareResource, CamelResource, RouteTemplateBeansAwareResource } from '../../camel';
import { CamelCatalogService, CatalogKind } from '../../index';
import { EntityType } from '../../camel/entities';
import { BeansEntity } from './index';
import { RouteTemplateBeansEntity } from './routeTemplateBeansEntity';
import {
  BeansDeserializer,
  RegistryBeanDefinition,
  RouteTemplateBeanDefinition,
} from '@kaoto-next/camel-catalog/types';
import { JSONSchemaType } from 'ajv';

/**
 * This class is to absorb a little bit of difference between beans such as {@link RegistryBeanDefinition} and {@link RouteTemplateBeanDefinition}.
 */
export class BeansEntityHandler {
  private type: 'beans' | 'routeTemplateBean' | undefined;
  private beansAware: BeansAwareResource | RouteTemplateBeansAwareResource | undefined;
  constructor(private camelResource?: CamelResource) {
    if (!this.camelResource) return;
    if ((this.camelResource as unknown as BeansAwareResource).createBeansEntity !== undefined) {
      this.beansAware = this.camelResource as unknown as BeansAwareResource;
      this.type = 'beans';
    } else if (
      (this.camelResource as unknown as RouteTemplateBeansAwareResource).createRouteTemplateBeansEntity !== undefined
    ) {
      this.beansAware = this.camelResource as unknown as RouteTemplateBeansAwareResource;
      this.type = 'routeTemplateBean';
    }
  }

  isSupported() {
    return this.type !== undefined;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  getBeanSchema(): JSONSchemaType<any> | undefined {
    switch (this.type) {
      case 'beans':
        return CamelCatalogService.getComponent(CatalogKind.Entity, 'bean')?.propertiesSchema;
      case 'routeTemplateBean':
        return CamelCatalogService.getComponent(CatalogKind.Entity, 'routeTemplateBean')?.propertiesSchema;
      default:
        return undefined;
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  getBeansSchema(): JSONSchemaType<any> | undefined {
    switch (this.type) {
      case 'beans': {
        const beanCatalog = CamelCatalogService.getComponent(CatalogKind.Entity, 'beans');
        return beanCatalog?.propertiesSchema;
      }
      case 'routeTemplateBean': {
        const beanCatalog = CamelCatalogService.getComponent(CatalogKind.Entity, 'routeTemplateBean');
        const schema = beanCatalog?.propertiesSchema;
        return !schema
          ? undefined
          : {
              title: schema.title,
              description: schema.description,
              type: 'array',
              definitions: schema.definitions,
              items: schema,
            };
      }
      default:
        return undefined;
    }
  }

  getBeansModel() {
    return this.getBeansEntity()?.parent.beans;
  }

  setBeansModel(model: BeansDeserializer | RouteTemplateBeanDefinition[]) {
    if (!Array.isArray(model) || model.length === 0) {
      const entity = this.getBeansEntity();
      entity && this.deleteBeansEntity(entity);
      return;
    }

    let entity = this.getBeansEntity();
    if (!entity) {
      entity = this.createBeansEntity();
    }
    if (entity) {
      entity.parent.beans = model;
    }
  }

  createBeansEntity() {
    switch (this.type) {
      case 'beans':
        return (this.beansAware as BeansAwareResource).createBeansEntity();
      case 'routeTemplateBean':
        return (this.beansAware as RouteTemplateBeansAwareResource).createRouteTemplateBeansEntity();
      default:
        return undefined;
    }
  }

  getBeansEntity() {
    switch (this.type) {
      case 'beans':
        return this.camelResource?.getEntities().find((item) => item.type === EntityType.Beans) as
          | BeansEntity
          | undefined;
      case 'routeTemplateBean':
        return (this.beansAware as RouteTemplateBeansAwareResource).getRouteTemplateBeansEntity();
      default:
        return undefined;
    }
  }

  deleteBeansEntity(entity: BeansEntity | RouteTemplateBeansEntity) {
    switch (this.type) {
      case 'beans':
        (this.beansAware as BeansAwareResource).deleteBeansEntity(entity as BeansEntity);
        break;
      case 'routeTemplateBean':
        (this.beansAware as RouteTemplateBeansAwareResource).deleteRouteTemplateBeansEntity();
        break;
    }
  }

  getAllBeansNameAndType(): { name: string; type: string }[] {
    const allBeans = this.getBeansModel();
    if (!allBeans) return [];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return allBeans?.map((bean: any) => {
      return { name: bean.name, type: bean.type };
    });
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  addNewBean(model: any) {
    const beansEntityToAdd = this.getBeansEntity() ? this.getBeansEntity() : this.createBeansEntity();
    beansEntityToAdd?.parent.beans.push(model);
  }

  getReferenceFromName(name: string) {
    switch (this.type) {
      case 'beans':
        return '#' + name;
      case 'routeTemplateBean':
        return `#bean:{{${name}}}`;
    }
  }

  getReferenceQuote() {
    switch (this.type) {
      case 'beans':
        return '#';
      case 'routeTemplateBean':
        return '#bean:{{}}';
    }
  }

  stripReferenceQuote(inputValue: string) {
    switch (this.type) {
      case 'beans':
        return inputValue && inputValue.startsWith('#') ? inputValue.substring(1) : inputValue;
      case 'routeTemplateBean': {
        const leftStripped = inputValue && inputValue.startsWith('#bean:{{') ? inputValue.substring(8) : inputValue;
        return leftStripped && leftStripped.endsWith('}}')
          ? leftStripped.substring(0, leftStripped.length - 2)
          : leftStripped;
      }
    }
  }
}
