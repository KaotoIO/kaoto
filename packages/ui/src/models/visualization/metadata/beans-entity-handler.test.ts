import catalogLibrary from '@kaoto/camel-catalog/index.json';
import { CatalogLibrary } from '@kaoto/camel-catalog/types';
import cloneDeep from 'lodash/cloneDeep';
import * as routeStub from '../../../stubs/camel-route';
import * as kameletStub from '../../../stubs/kamelet-route';
import { getFirstCatalogMap } from '../../../stubs/test-load-catalog';
import { CamelRouteResource, KameletResource, PipeResource } from '../../camel';
import { CatalogKind } from '../../catalog-kind';
import { KaotoSchemaDefinition } from '../../kaoto-schema';
import { CamelCatalogService } from '../flows';
import { BeansEntityHandler } from './beans-entity-handler';

describe('BeansEntityHandler', () => {
  beforeAll(async () => {
    const catalogsMap = await getFirstCatalogMap(catalogLibrary as CatalogLibrary);
    CamelCatalogService.setCatalogKey(CatalogKind.Entity, catalogsMap.entitiesCatalog);
  });

  describe('should handle beans in CamelRouteResource', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let model: any;
    let beansHandler: BeansEntityHandler;
    beforeEach(() => {
      model = cloneDeep(routeStub.camelRouteJson);
      const camelRouteResource = new CamelRouteResource(model);
      beansHandler = new BeansEntityHandler(camelRouteResource);
    });

    it('initial state', () => {
      expect(model.beans).toBeUndefined();
      expect(beansHandler.isSupported()).toBeTruthy();
      const beanSchema = beansHandler.getBeanSchema();
      expect(beanSchema!.properties!.builderClass.title).toEqual('Builder Class');
      expect(beanSchema!.properties!.script.title).toEqual('Script');
      const beansSchema = beansHandler.getBeansSchema();
      expect((beansSchema!.items as KaotoSchemaDefinition['schema'])['$ref']).toContain('BeanFactoryDefinition');
      expect(beansHandler.getBeansEntity()).toBeUndefined();
      expect(beansHandler.getBeansModel()).toBeUndefined();
    });

    it('reference quote', () => {
      expect(beansHandler.getReferenceQuote()).toEqual('#');
      expect(beansHandler.getReferenceFromName('myBean')).toEqual('#myBean');
      expect(beansHandler.stripReferenceQuote('#myBean')).toEqual('myBean');
    });

    it('create and delete bean entity', () => {
      beansHandler.createBeansEntity();
      expect(beansHandler.getBeansEntity()?.type).toEqual('beans');
      expect(beansHandler.getBeansModel()).toEqual([]);
      beansHandler.setBeansModel([
        { name: 'myBean1', type: 'myType1', properties: { p: 'v' } },
        { name: 'myBean2', type: 'myType2', properties: { p: 'v' } },
      ]);
      expect(beansHandler.getBeansModel()?.length).toEqual(2);
      const nameAndType = beansHandler.getAllBeansNameAndType();
      expect(nameAndType.length).toEqual(2);
      expect(nameAndType).toEqual([
        { name: 'myBean1', type: 'myType1' },
        { name: 'myBean2', type: 'myType2' },
      ]);
      beansHandler.addNewBean({ name: 'myBean3', type: 'myType3' });
      expect(beansHandler.getBeansModel()?.length).toEqual(3);
      const entity = beansHandler.getBeansEntity();
      expect(entity).toBeDefined();
      beansHandler.deleteBeansEntity(entity!);
      expect(beansHandler.getBeansEntity()).toBeUndefined();
      expect(beansHandler.getBeansModel()).toBeUndefined();
    });
  });

  describe('should handle beans in KameletResource', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let model: any;
    let beansHandler: BeansEntityHandler;
    beforeEach(() => {
      model = cloneDeep(kameletStub.kameletJson);
      const kameletResource = new KameletResource(model);
      beansHandler = new BeansEntityHandler(kameletResource);
    });

    it('initial state', () => {
      expect(model.spec.template.beans).toBeUndefined();
      expect(beansHandler.isSupported()).toBeTruthy();
      const beanSchema = beansHandler.getBeanSchema();
      expect(beanSchema?.properties!.builderClass).toBeDefined();
      expect(beanSchema?.properties!.script.title).toEqual('Script');
      const beansSchema = beansHandler.getBeansSchema();
      expect((beansSchema?.items as KaotoSchemaDefinition['schema']).properties!.scriptLanguage.title).toEqual(
        'Script Language',
      );
      expect((beansSchema?.items as KaotoSchemaDefinition['schema']).properties!.builderClass).toBeDefined();
      expect(beansHandler.getBeansEntity()).toBeUndefined();
      expect(beansHandler.getBeansModel()).toBeUndefined();
    });

    it('reference quote', () => {
      expect(beansHandler.getReferenceQuote()).toEqual('#bean:{{}}');
      expect(beansHandler.getReferenceFromName('myBean')).toEqual('#bean:{{myBean}}');
      expect(beansHandler.stripReferenceQuote('#bean:{{myBean}}')).toEqual('myBean');
    });

    it('create and delete bean entity', () => {
      beansHandler.createBeansEntity();
      expect(beansHandler.getBeansEntity()?.type).toEqual('beans');
      expect(beansHandler.getBeansModel()).toEqual([]);
      beansHandler.setBeansModel([
        { name: 'myBean1', type: 'myType1', properties: { p: 'v' } },
        { name: 'myBean2', type: 'myType2', properties: { p: 'v' } },
      ]);
      expect(beansHandler.getBeansModel()?.length).toEqual(2);
      const nameAndType = beansHandler.getAllBeansNameAndType();
      expect(nameAndType.length).toEqual(2);
      expect(nameAndType).toEqual([
        { name: 'myBean1', type: 'myType1' },
        { name: 'myBean2', type: 'myType2' },
      ]);
      beansHandler.addNewBean({ name: 'myBean3', type: 'myType3' });
      expect(beansHandler.getBeansModel()?.length).toEqual(3);
      const entity = beansHandler.getBeansEntity();
      expect(entity).toBeDefined();
      beansHandler.deleteBeansEntity(entity!);
      expect(beansHandler.getBeansEntity()).toBeUndefined();
      expect(beansHandler.getBeansModel()).toBeUndefined();
    });
  });

  describe('should handle erroneous CamelResource', () => {
    it('if CamelResource is undefined', () => {
      const beansHandler = new BeansEntityHandler();
      expect(beansHandler.isSupported()).toBeFalsy();
      expect(beansHandler.stripReferenceQuote('#myBean')).toBeUndefined();
      beansHandler.addNewBean({ name: 'myBean', type: 'myType' });
      expect(beansHandler.getBeansEntity()).toBeUndefined();
      beansHandler.setBeansModel([{ name: 'myBean', type: 'myType' }]);
      expect(beansHandler.getReferenceFromName('myBean')).toBeUndefined();
      expect(beansHandler.stripReferenceQuote('#myBean')).toBeUndefined();
    });

    it('if CamelResource is not the supported one', () => {
      const camelResource = new PipeResource({});
      const beansHandler = new BeansEntityHandler(camelResource);
      expect(beansHandler.isSupported()).toBeFalsy();
      expect(beansHandler.stripReferenceQuote('#myBean')).toBeUndefined();
      beansHandler.addNewBean({ name: 'myBean', type: 'myType' });
      expect(beansHandler.getBeansEntity()).toBeUndefined();
      beansHandler.setBeansModel([{ name: 'myBean', type: 'myType' }]);
      expect(beansHandler.getReferenceFromName('myBean')).toBeUndefined();
      expect(beansHandler.stripReferenceQuote('#myBean')).toBeUndefined();
    });
  });
});
