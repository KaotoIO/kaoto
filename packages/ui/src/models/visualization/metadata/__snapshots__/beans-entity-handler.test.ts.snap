// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeansEntityHandler should handle beans in CamelRouteResource initial state 1`] = `
{
  "additionalProperties": false,
  "description": "Define custom beans that can be used in your Camel routes and in general.",
  "properties": {
    "builderClass": {
      "$comment": "group:common",
      "description": "Fully qualified class name of builder class to use for creating and configuring the bean. The builder will use the properties values to configure the bean.",
      "title": "Builder Class",
      "type": "string",
    },
    "builderMethod": {
      "$comment": "group:common",
      "default": "build",
      "description": "Name of method when using builder class. This method is invoked after configuring to create the actual bean. This method is often named build (used by default).",
      "title": "Builder Method",
      "type": "string",
    },
    "constructors": {
      "$comment": "group:common",
      "description": "Optional constructor arguments for creating the bean. Arguments correspond to specific index of the constructor argument list, starting from zero.",
      "title": "Constructors",
      "type": "object",
    },
    "destroyMethod": {
      "$comment": "group:common",
      "description": "The name of the custom destroy method to invoke on bean shutdown, such as when Camel is shutting down. The method must have no arguments, but may throw any exception.",
      "title": "Destroy Method",
      "type": "string",
    },
    "factoryBean": {
      "$comment": "group:common",
      "description": "Name of factory bean (bean id) to use for creating the bean.",
      "title": "Factory Bean",
      "type": "string",
    },
    "factoryMethod": {
      "$comment": "group:common",
      "description": "Name of method to invoke when creating the bean via a factory bean.",
      "title": "Factory Method",
      "type": "string",
    },
    "initMethod": {
      "$comment": "group:common",
      "description": "The name of the custom initialization method to invoke after setting bean properties. The method must have no arguments, but may throw any exception.",
      "title": "Init Method",
      "type": "string",
    },
    "name": {
      "$comment": "group:common",
      "description": "The name of the bean (bean id)",
      "title": "Name",
      "type": "string",
    },
    "properties": {
      "$comment": "group:common",
      "description": "Optional properties to set on the created bean.",
      "title": "Properties",
      "type": "object",
    },
    "script": {
      "$comment": "group:advanced",
      "description": "The script to execute that creates the bean when using scripting languages. If the script use the prefix resource: such as resource:classpath:com/foo/myscript.groovy, resource:file:/var/myscript.groovy, then its loaded from the external resource.",
      "title": "Script",
      "type": "string",
    },
    "scriptLanguage": {
      "$comment": "group:advanced",
      "description": "The script language to use when using inlined script for creating the bean, such as groovy, java, javascript etc.",
      "title": "Script Language",
      "type": "string",
    },
    "type": {
      "$comment": "group:common",
      "description": "The class name (fully qualified) of the bean",
      "title": "Type",
      "type": "string",
    },
  },
  "required": [
    "name",
    "type",
  ],
  "title": "Bean Factory",
  "type": "object",
}
`;
